### ISSUE TITLE
Front Page DeepExplainer MNIST tutorial is broken

### ISSUE THREAD

**Issue by palatyle** [shap/shap#1343]
Hi all,

I'm trying to run the DeepExplainer MNIST tutorial, but am running into errors. I've tried this on both Mac and Windows with the same errors. Everything works up until this line:

`shap_values = e.shap_values(x_test[1:5])`

Where I get the following string of errors: 

```
InvalidArgumentError                      Traceback (most recent call last)
~/opt/anaconda3/lib/python3.7/site-packages/tensorflow/python/framework/ops.py in get_attr(self, name)
   2638       with c_api_util.tf_buffer() as buf:
-> 2639         c_api.TF_OperationGetAttrValueProto(self._c_op, name, buf)
   2640         data = c_api.TF_GetBuffer(buf)

InvalidArgumentError: Operation 'dense_2/Softmax' has no attr named '_XlaCompile'.

During handling of the above exception, another exception occurred:

ValueError                                Traceback (most recent call last)
~/opt/anaconda3/lib/python3.7/site-packages/tensorflow/python/ops/gradients_util.py in _MaybeCompile(scope, op, func, grad_fn)
    397     try:
--> 398       xla_compile = op.get_attr("_XlaCompile")
    399       xla_separate_compiled_gradients = op.get_attr(

~/opt/anaconda3/lib/python3.7/site-packages/tensorflow/python/framework/ops.py in get_attr(self, name)
   2642       # Convert to ValueError for backwards compatibility.
-> 2643       raise ValueError(str(e))
   2644     x = attr_value_pb2.AttrValue()

ValueError: Operation 'dense_2/Softmax' has no attr named '_XlaCompile'.

During handling of the above exception, another exception occurred:

InvalidArgumentError                      Traceback (most recent call last)
~/opt/anaconda3/lib/python3.7/site-packages/tensorflow/python/framework/ops.py in get_attr(self, name)
   2638       with c_api_util.tf_buffer() as buf:
-> 2639         c_api.TF_OperationGetAttrValueProto(self._c_op, name, buf)
   2640         data = c_api.TF_GetBuffer(buf)

InvalidArgumentError: Operation 'gradients_1/dense_2/Softmax_grad/truediv' has no attr named '_XlaCompile'.

During handling of the above exception, another exception occurred:

ValueError                                Traceback (most recent call last)
~/opt/anaconda3/lib/python3.7/site-packages/tensorflow/python/ops/gradients_util.py in _MaybeCompile(scope, op, func, grad_fn)
    397     try:
--> 398       xla_compile = op.get_attr("_XlaCompile")
    399       xla_separate_compiled_gradients = op.get_attr(

~/opt/anaconda3/lib/python3.7/site-packages/tensorflow/python/framework/ops.py in get_attr(self, name)
   2642       # Convert to ValueError for backwards compatibility.
-> 2643       raise ValueError(str(e))
   2644     x = attr_value_pb2.AttrValue()

ValueError: Operation 'gradients_1/dense_2/Softmax_grad/truediv' has no attr named '_XlaCompile'.

During handling of the above exception, another exception occurred:

InvalidArgumentError                      Traceback (most recent call last)
~/opt/anaconda3/lib/python3.7/site-packages/tensorflow/python/framework/ops.py in _create_c_op(graph, node_def, inputs, control_inputs)
   1863   try:
-> 1864     c_op = c_api.TF_FinishOperation(op_desc)
   1865   except errors.InvalidArgumentError as e:

InvalidArgumentError: Dimension 1 in both shapes must be equal, but are 10 and 1. Shapes are [?,10] and [?,1]. for 'gradients_1/dense_2/Softmax_grad/gradients/gradients_1/dense_2/Softmax_grad/truediv_grad/Select_1' (op: 'Select') with input shapes: [?,1], [?,10], [?,10].

During handling of the above exception, another exception occurred:

ValueError                                Traceback (most recent call last)
<ipython-input-5-f9a2d473c781> in <module>
      1 # ...or pass tensors directly
      2 # e = shap.DeepExplainer((model.layers[0].input, model.layers[-1].output), background)
----> 3 shap_values = e.shap_values(x_test[1:5])

~/opt/anaconda3/lib/python3.7/site-packages/shap/explainers/deep/__init__.py in shap_values(self, X, ranked_outputs, output_rank_order, check_additivity)
    117         were chosen as "top".
    118         """
--> 119         return self.explainer.shap_values(X, ranked_outputs, output_rank_order, check_additivity=check_additivity)

~/opt/anaconda3/lib/python3.7/site-packages/shap/explainers/deep/deep_tf.py in shap_values(self, X, ranked_outputs, output_rank_order, check_additivity)
    302                 # run attribution computation graph
    303                 feature_ind = model_output_ranks[j,i]
--> 304                 sample_phis = self.run(self.phi_symbolic(feature_ind), self.model_inputs, joint_input)
    305 
    306                 # assign the attributions to the right part of the output arrays

~/opt/anaconda3/lib/python3.7/site-packages/shap/explainers/deep/deep_tf.py in phi_symbolic(self, i)
    228                     return tf.gradients(out, self.model_inputs)
    229 
--> 230                 self.phi_symbolics[i] = self.execute_with_overridden_gradients(anon)
    231             else:
    232                 @tf.function

~/opt/anaconda3/lib/python3.7/site-packages/shap/explainers/deep/deep_tf.py in execute_with_overridden_gradients(self, f)
    395         # define the computation graph for the attribution values using a custom gradient-like computation
    396         try:
--> 397             out = f()
    398         finally:
    399             # reinstate the backpropagatable check

~/opt/anaconda3/lib/python3.7/site-packages/shap/explainers/deep/deep_tf.py in anon()
    226                 def anon():
    227                     out = self.model_output[:,i] if self.multi_output else self.model_output
--> 228                     return tf.gradients(out, self.model_inputs)
    229 
    230                 self.phi_symbolics[i] = self.execute_with_overridden_gradients(anon)

~/opt/anaconda3/lib/python3.7/site-packages/tensorflow/python/ops/gradients_impl.py in gradients(ys, xs, grad_ys, name, colocate_gradients_with_ops, gate_gradients, aggregation_method, stop_gradients, unconnected_gradients)
    156         ys, xs, grad_ys, name, colocate_gradients_with_ops,
    157         gate_gradients, aggregation_method, stop_gradients,
--> 158         unconnected_gradients)
    159   # pylint: enable=protected-access
    160 

~/opt/anaconda3/lib/python3.7/site-packages/tensorflow/python/ops/gradients_util.py in _GradientsHelper(ys, xs, grad_ys, name, colocate_gradients_with_ops, gate_gradients, aggregation_method, stop_gradients, unconnected_gradients, src_graph)
    729                 # functions.
    730                 in_grads = _MaybeCompile(grad_scope, op, func_call,
--> 731                                          lambda: grad_fn(op, *out_grads))
    732               else:
    733                 # For function call ops, we add a 'SymbolicGradient'

~/opt/anaconda3/lib/python3.7/site-packages/tensorflow/python/ops/gradients_util.py in _MaybeCompile(scope, op, func, grad_fn)
    401       xla_scope = op.get_attr("_XlaScope").decode()
    402     except ValueError:
--> 403       return grad_fn()  # Exit early
    404 
    405   if not xla_compile:

~/opt/anaconda3/lib/python3.7/site-packages/tensorflow/python/ops/gradients_util.py in <lambda>()
    729                 # functions.
    730                 in_grads = _MaybeCompile(grad_scope, op, func_call,
--> 731                                          lambda: grad_fn(op, *out_grads))
    732               else:
    733                 # For function call ops, we add a 'SymbolicGradient'

~/opt/anaconda3/lib/python3.7/site-packages/shap/explainers/deep/deep_tf.py in custom_grad(self, op, *grads)
    365         """
    366         type_name = op.type[5:] if op.type.startswith("shap_") else op.type
--> 367         out = op_handlers[type_name](self, op, *grads) # we cut off the shap_ prefex before the lookup
    368         return out
    369 

~/opt/anaconda3/lib/python3.7/site-packages/shap/explainers/deep/deep_tf.py in softmax(explainer, op, *grads)
    480                 explainer.between_tensors[t.name] = False
    481 
--> 482     out = tf.gradients(div, in0_centered, grad_ys=grads[0])[0]
    483 
    484     # remove the names we just added

~/opt/anaconda3/lib/python3.7/site-packages/tensorflow/python/ops/gradients_impl.py in gradients(ys, xs, grad_ys, name, colocate_gradients_with_ops, gate_gradients, aggregation_method, stop_gradients, unconnected_gradients)
    156         ys, xs, grad_ys, name, colocate_gradients_with_ops,
    157         gate_gradients, aggregation_method, stop_gradients,
--> 158         unconnected_gradients)
    159   # pylint: enable=protected-access
    160 

~/opt/anaconda3/lib/python3.7/site-packages/tensorflow/python/ops/gradients_util.py in _GradientsHelper(ys, xs, grad_ys, name, colocate_gradients_with_ops, gate_gradients, aggregation_method, stop_gradients, unconnected_gradients, src_graph)
    729                 # functions.
    730                 in_grads = _MaybeCompile(grad_scope, op, func_call,
--> 731                                          lambda: grad_fn(op, *out_grads))
    732               else:
    733                 # For function call ops, we add a 'SymbolicGradient'

~/opt/anaconda3/lib/python3.7/site-packages/tensorflow/python/ops/gradients_util.py in _MaybeCompile(scope, op, func, grad_fn)
    401       xla_scope = op.get_attr("_XlaScope").decode()
    402     except ValueError:
--> 403       return grad_fn()  # Exit early
    404 
    405   if not xla_compile:

~/opt/anaconda3/lib/python3.7/site-packages/tensorflow/python/ops/gradients_util.py in <lambda>()
    729                 # functions.
    730                 in_grads = _MaybeCompile(grad_scope, op, func_call,
--> 731                                          lambda: grad_fn(op, *out_grads))
    732               else:
    733                 # For function call ops, we add a 'SymbolicGradient'

~/opt/anaconda3/lib/python3.7/site-packages/shap/explainers/deep/deep_tf.py in custom_grad(self, op, *grads)
    365         """
    366         type_name = op.type[5:] if op.type.startswith("shap_") else op.type
--> 367         out = op_handlers[type_name](self, op, *grads) # we cut off the shap_ prefex before the lookup
    368         return out
    369 

~/opt/anaconda3/lib/python3.7/site-packages/shap/explainers/deep/deep_tf.py in handler(explainer, op, *grads)
    555             return linearity_1d_handler(input_ind1, explainer, op, *grads)
    556         elif var[input_ind0] and var[input_ind1]:
--> 557             return nonlinearity_2d_handler(input_ind0, input_ind1, op_func, explainer, op, *grads)
    558         else:
    559             return [None for _ in op.inputs] # no inputs vary, we must be hidden by a switch function

~/opt/anaconda3/lib/python3.7/site-packages/shap/explainers/deep/deep_tf.py in nonlinearity_2d_handler(input_ind0, input_ind1, op_func, explainer, op, *grads)
    626     # Avoid divide by zero nans
    627     out0 = tf.where(tf.abs(tf.tile(delta_in0, dup0)) < 1e-7, tf.zeros_like(out0), out0)
--> 628     out1 = tf.where(tf.abs(tf.tile(delta_in1, dup0)) < 1e-7, tf.zeros_like(out1), out1)
    629 
    630     # see if due to broadcasting our gradient shapes don't match our input shapes

~/opt/anaconda3/lib/python3.7/site-packages/tensorflow/python/util/deprecation.py in new_func(*args, **kwargs)
    322               'in a future version' if date is None else ('after %s' % date),
    323               instructions)
--> 324       return func(*args, **kwargs)
    325     return tf_decorator.make_decorator(
    326         func, new_func, 'deprecated',

~/opt/anaconda3/lib/python3.7/site-packages/tensorflow/python/util/dispatch.py in wrapper(*args, **kwargs)
    178     """Call target, and fall back on dispatchers if there is a TypeError."""
    179     try:
--> 180       return target(*args, **kwargs)
    181     except (TypeError, ValueError):
    182       # Note: convert_to_eager_tensor currently raises a ValueError, not a

~/opt/anaconda3/lib/python3.7/site-packages/tensorflow/python/ops/array_ops.py in where(condition, x, y, name)
   3268       return gen_array_ops.where(condition=condition, name=name)
   3269   elif x is not None and y is not None:
-> 3270     return gen_math_ops.select(condition=condition, x=x, y=y, name=name)
   3271   else:
   3272     raise ValueError("x and y must both be non-None or both be None.")

~/opt/anaconda3/lib/python3.7/site-packages/tensorflow/python/ops/gen_math_ops.py in select(condition, x, y, name)
   9224   # Add nodes to the TensorFlow graph.
   9225   _, _, _op = _op_def_lib._apply_op_helper(
-> 9226         "Select", condition=condition, t=x, e=y, name=name)
   9227   _result = _op.outputs[:]
   9228   _inputs_flat = _op.inputs

~/opt/anaconda3/lib/python3.7/site-packages/tensorflow/python/framework/op_def_library.py in _apply_op_helper(self, op_type_name, name, **keywords)
    786         op = g.create_op(op_type_name, inputs, dtypes=None, name=scope,
    787                          input_types=input_types, attrs=attr_protos,
--> 788                          op_def=op_def)
    789       return output_structure, op_def.is_stateful, op
    790 

~/opt/anaconda3/lib/python3.7/site-packages/tensorflow/python/util/deprecation.py in new_func(*args, **kwargs)
    505                 'in a future version' if date is None else ('after %s' % date),
    506                 instructions)
--> 507       return func(*args, **kwargs)
    508 
    509     doc = _add_deprecated_arg_notice_to_docstring(

~/opt/anaconda3/lib/python3.7/site-packages/tensorflow/python/framework/ops.py in create_op(***failed resolving arguments***)
   3614           input_types=input_types,
   3615           original_op=self._default_original_op,
-> 3616           op_def=op_def)
   3617       self._create_op_helper(ret, compute_device=compute_device)
   3618     return ret

~/opt/anaconda3/lib/python3.7/site-packages/tensorflow/python/framework/ops.py in __init__(self, node_def, g, inputs, output_types, control_inputs, input_types, original_op, op_def)
   2025           op_def, inputs, node_def.attr)
   2026       self._c_op = _create_c_op(self._graph, node_def, grouped_inputs,
-> 2027                                 control_input_ops)
   2028 
   2029     # Initialize self._outputs.

~/opt/anaconda3/lib/python3.7/site-packages/tensorflow/python/framework/ops.py in _create_c_op(graph, node_def, inputs, control_inputs)
   1865   except errors.InvalidArgumentError as e:
   1866     # Convert to ValueError for backwards compatibility.
-> 1867     raise ValueError(str(e))
   1868 
   1869   return c_op

ValueError: Dimension 1 in both shapes must be equal, but are 10 and 1. Shapes are [?,10] and [?,1]. for 'gradients_1/dense_2/Softmax_grad/gradients/gradients_1/dense_2/Softmax_grad/truediv_grad/Select_1' (op: 'Select') with input shapes: [?,1], [?,10], [?,10].
```

Any ideas? I'm on TensorFlow 1.14.0, Keras 2.3.1, and shap 0.35.0

**Comment by slundberg** [comment]
When I run the notebook using my TF 2.3 install I don't get any errors, but it should also work on TF 1.x...

**Comment by nocoffeenotea** [comment]
it is work when shap==0.36.0
tf==2.3

**Comment by niloofardadras** [comment]
hey, I am also getting same error for deep explainer and keras sequential model. also tried shap==0.36.0
tf==2.3 and nothing changed. the strange thing is that it happens sometimes when i call the explainer multiple times. any help is appreciated.

**Comment by connortann** [comment]
This should be fixed by #3393
